jgrapht-core/src/main/java/org/jgrapht/alg/connectivity/DulmageMendelsohnDecomposition.java

```diff
@@ -0,0 +1,303 @@
+        return decompose(matching, fine);
+    }
+
+    /**
+     * Perform the decomposition, using a precalculated bipartite matching
+     *
+     * @param matching the matching from a MatchingAlgorithm
+     * @param fine true if the fine decomposition is required
+     * @return the Decomposition
+     */
+    public Decomposition<V, E> decompose(Matching<V, E> matching, boolean fine) {
+        // Determine the unmatched vertices from both partitions
+        Set<V> unmatched1 = new HashSet<>();
+        Set<V> unmatched2 = new HashSet<>();
+        getUnmatched(matching, unmatched1, unmatched2);
+        // Assemble a directed graph
+        Graph<V, DefaultEdge> dg = asDirectedGraph(matching);
+        // Find the non-square subgraph dominated by partition1
+        Set<V> subset1 = new HashSet<>();
+        unmatched1.stream().map((v) -> {
+            subset1.add(v);
+            return v;
+        }).map((v) -> new DepthFirstIterator<>(dg, v)).forEachOrdered((it) -> {
+            while (it.hasNext()) {
+                subset1.add(it.next());
+            }
+        });
+        // Find the non-square subgraph dominated by partition2
+        Graph<V, DefaultEdge> gd = new EdgeReversedGraph<>(dg);
+        Set<V> subset2 = new HashSet<>();
+        unmatched2.stream().map((v) -> {
+            subset2.add(v);
+            return v;
+        }).map((v) -> new DepthFirstIterator<>(gd, v)).forEachOrdered((it) -> {
+            while (it.hasNext()) {
+                subset2.add(it.next());
+            }
+        });
+        // Find the square subgraph
+        Set<V> subset3 = new HashSet<>();
+        subset3.addAll(partition1);
+        subset3.addAll(partition2);
+        subset3.removeAll(subset1);
+        subset3.removeAll(subset2);
+        if (fine) {
+            List<Set<V>> out = new ArrayList<>();
+            // Build a directed graph between edges of the matching in subset3
+            Graph<E, DefaultEdge> H = asDirectedEdgeGraph(matching, subset3);
+
+            // Perform strongly-connected-components on the graph
+            StrongConnectivityAlgorithm<E, DefaultEdge> sci
+                    = new KosarajuStrongConnectivityInspector<>(H);
+            // Divide into sets of vertices
+            for (Set<E> edgeSet : sci.stronglyConnectedSets()) {
+                Set<V> vertexSet = new HashSet<>();
+                edgeSet.stream().map((edge) -> {
+                    vertexSet.add(graph.getEdgeSource(edge));
+                    return edge;
+                }).forEachOrdered((edge) -> {
+                    vertexSet.add(graph.getEdgeTarget(edge));
+                });
+                out.add(vertexSet);
+            }
+            return new Decomposition<>(graph, subset1, subset2, out);
+        } else {
+            return new Decomposition<>(graph, subset1, subset2, Arrays.asList(subset3));
+        }
+    }
+
+    /**
+     * The output of a decomposition operation
```

Comment:
Can you refine the description a bit. What exactly are these subsets. Please related to the paper, and/or the wiki article. I'm not confident that a user understands what `getPartition1DominatedSet` or `getPartition2DominatedSet` returns based on the current (limited) description. It should generally not be necessary for a user to look into the actual implementation to understand what's being returned.
Consider adopting the notation of the wiki article https://en.wikipedia.org/wiki/Dulmage%E2%80%93Mendelsohn_decomposition#The_coarse_decomposition
or the notation used in the paper used for the implementation of this class